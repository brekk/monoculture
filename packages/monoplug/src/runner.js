import {
  curry,
  pipe,
  reduce,
  identity as I,
  mergeRight,
  prop,
  map,
  fromPairs,
} from 'ramda'
import { topologicalDependencySort } from './sort'

export const taskProcessor = curry((context, plugins) =>
  pipe(
    topologicalDependencySort,
    reduce(
      (
        { state, events },
        { name, fn, selector = I, store: __focus = false }
      ) => {
        const store = __focus || I
        const outcome = fn(selector(state))
        const newState = { ...state, [name]: outcome }
        return {
          events: [...events, [name, outcome]],
          state: __focus ? store(newState) : newState,
        }
      },
      { state: context, events: [] }
    )
  )(plugins)
)

const stepFunction = curry((selected, { processLine, fn }, file) =>
  processLine
    ? {
        ...file,
        body: map(([k, v]) => [k, fn(selected, v)])(file.body),
      }
    : fn(selected, file)
)

const processRelativeToFile = curry((state, plugin, files) =>
  reduce(
    (agg, __file) => {
      const {
        name,
        fn,
        selector = I,
        store: __focus = false,
        processLine = false,
      } = plugin
      const selected = selector(state)
      const { hash } = __file
      console.log('processLine', processLine, fn)

      // I figure we wanna keep things as arrays for lookups
      // but might entirely change this with respect to futures over time
      return [...agg, [hash, stepFunction(selected, plugin, __file)]]
    },
    [],
    files
  )
)

// files: List { file :: string, body :: List [Number, String] }
export const fileProcessor = curry((context, plugins, files) =>
  pipe(
    topologicalDependencySort,
    reduce(
      (
        // context
        { state, events },
        // plugin
        plugin
      ) => {
        const { store: __focus = false, name } = plugin
        const store = __focus || I
        const outcome = processRelativeToFile(state, plugin, files)
        const newState = { ...state, [name]: outcome }
        return {
          events: [...events, name],
          state: __focus ? store(newState) : newState,
        }
      },
      { state: context, events: [] }
    ),
    mergeRight({
      hashMap: pipe(
        map(x => [prop('hash', x), prop('file', x)]),
        fromPairs
      )(files),
    })
  )(plugins)
)
