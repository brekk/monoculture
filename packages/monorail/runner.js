import {
  curry,
  fromPairs,
  tap,
  identity as I,
  map,
  pipe,
  prop,
  reduce,
} from 'ramda'
import { pap, resolve } from 'fluture'
import { makeFileHelpers, makePluginHelpers } from './helpers'
import { toposort } from './sort'
import { log } from './log'

export const stepFunction = curry(function _stepFunction(state, plugin, file) {
  tap(({ name, dependencies: d }) => log.run(`plugin [${name}]`, d))
  tap(({ name, hash }) => log.run(`file [${name}]`, hash))
  const { selector = I, preserveLine = false, fn } = plugin
  const selected = selector(state)
  const base = makeFileHelpers(file)
  const plugged = makePluginHelpers(state, plugin, log)
  const helpers = { ...base, ...plugged }
  const output = preserveLine
    ? {
        ...file,
        body: map(([k, v]) => [k, fn(selected, v, helpers)])(file.body),
      }
    : fn(selected, file, helpers)
  return output
})

const runPluginOnFilesWithContext = curry(
  function _runPluginOnFilesWithContext(context, files, plugin) {
    if (!plugin.name) return []
    log.run('applying plugin', plugin.name)
    return pipe(
      map(file => [file.name, stepFunction(context, plugin, file)]),
      fromPairs
    )(files)
  }
)

export const getNames = map(prop('name'))
export const getHashes = pipe(
  map(z => [prop('name', z), prop('hash', z)]),
  fromPairs
)

export const statefulApplicator = curry(
  function _statefulApplicator(context, plugins, files) {
    // drop info we don't want to retain downstream
    const { HELP: _h, basePath: _b, cwd: _c, ...config } = context
    return reduce(
      (agg, plugin) => ({
        ...agg,
        state: {
          ...agg.state,
          [plugin.name]: pipe(runPluginOnFilesWithContext(agg, files))(plugin),
        },
      }),
      {
        state: {},
        filenames: getNames(files),
        hashes: getHashes(files),
        plugins: getNames(plugins),
        config,
      },
      plugins
    )
  }
)

export const futureFileProcessor = curry((context, pluginsF, filesF) =>
  pipe(
    resolve,
    pap(map(toposort, pluginsF)),
    pap(filesF)
  )(statefulApplicator(context))
)
