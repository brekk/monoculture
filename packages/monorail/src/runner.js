import {
  curry,
  fromPairs,
  identity as I,
  map,
  mergeRight,
  pipe,
  prop,
  reduce,
} from 'ramda'
import { pap, resolve } from 'fluture'
import { topologicalDependencySort } from './sort'

export const taskProcessor = curry((context, plugins) =>
  pipe(
    topologicalDependencySort,
    reduce(
      (
        { state, events },
        { name, fn, selector = I, store: __focus = false }
      ) => {
        const store = __focus || I
        const outcome = fn(selector(state))
        const newState = { ...state, [name]: outcome }
        return {
          events: [...events, [name, outcome]],
          state: __focus ? store(newState) : newState,
        }
      },
      { state: context, events: [] }
    )
  )(plugins)
)

export const stepFunction = curry(
  (state, { selector = I, preserveLine = false, fn }, file) => {
    const selected = selector(state)
    return preserveLine
      ? {
          ...file,
          body: map(([k, v]) => [k, fn(selected, v)])(file.body),
        }
      : fn(selected, file)
  }
)

const processRelativeToFile = curry((state, plugin, files) =>
  reduce(
    (agg, __file) => {
      const { hash } = __file
      return [...agg, [hash, stepFunction(state, plugin, __file)]]
    },
    [],
    files
  )
)

// files: List { file :: string, body :: List [Number, String] }
export const fileProcessor = curry((context, plugins, files) =>
  pipe(
    topologicalDependencySort,
    reduce(
      (
        // context
        { state, events },
        // plugin
        plugin
      ) => {
        const { store = I, name } = plugin
        const outcome = processRelativeToFile(state, plugin, files)
        const newState = { ...state, [name]: outcome }
        return {
          events: [...events, name],
          state: store(newState),
        }
      },
      { state: context, events: [] }
    ),
    mergeRight({
      hashMap: pipe(
        map(x => [prop('hash', x), prop('file', x)]),
        fromPairs
      )(files),
    })
  )(plugins)
)

export const futureApplicator = curry((context, plugins, files) => ({
  state: pipe(
    map(plugin => [
      plugin.name,
      pipe(
        map(file => [file.hash, stepFunction(context, plugin, file)]),
        fromPairs
      )(files),
    ]),
    fromPairs
  )(plugins),
  files,
  hashes: pipe(
    map(z => [prop('hash', z), prop('file', z)]),
    fromPairs
  )(files),
  plugins: map(prop('name'), plugins),
}))

export const futureFileProcessor = curry((context, pluginsF, filesF) =>
  pipe(
    pap(map(topologicalDependencySort, pluginsF)),
    pap(filesF)
  )(resolve(futureApplicator(context)))
)
