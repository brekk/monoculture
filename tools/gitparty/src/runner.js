import path from 'node:path'
import { parse as parseTime } from 'date-fns'
import {
  zonedTimeToUtc,
  utcToZonedTime,
  format as formatDate,
} from 'date-fns-tz'
import { applyPatternsWithChalk } from './per-commit'
import { printLegend } from './legend'
import { trace } from 'xtrace'
import { configFileWithCancel, configurate } from 'climate'
import { Chalk } from 'chalk'
import {
  swap,
  mapRej,
  resolve,
  coalesce,
  reject as rejectF,
  ap as apF,
  and as andF,
} from 'fluture'
import { writeFileWithConfigAndCancel, findUpWithCancel } from 'file-system'
import {
  F,
  T,
  __ as $,
  always as K,
  ap,
  join,
  chain,
  cond,
  curry,
  fromPairs,
  groupBy,
  head,
  identity as I,
  includes,
  map,
  mergeRight,
  objOf,
  pipe,
  prop,
  propOr,
  reject,
  split,
  startsWith,
  toPairs,
  uniq,
  zip,
} from 'ramda'
import { canonicalize } from './alias'
import {
  THIS_IS_NOT_A_GIT_REPO,
  CONFIG_NOT_VALID,
  GITPARTY_CONFIG_NEEDED,
} from './errors'
import {
  DEFAULT_CONFIG_FILE,
  YARGS_CONFIG,
  CONFIG_DEFAULTS,
  HELP_CONFIG,
} from './config'
import { gitlogWithCancel } from './git'
import { log } from './log'
import PKG from '../package.json'

const j2 = x => JSON.stringify(x, null, 2)

const { name: $NAME, description: $DESC } = PKG

const writeInitConfigFileWithCancel = curry((cancel, filepath) =>
  pipe(
    j2,
    writeFileWithConfigAndCancel(cancel, { encoding: 'utf8' }, filepath),
    map(K(`Wrote file to "${filepath}"!`))
  )(DEFAULT_CONFIG_FILE)
)

const loadPartyFile = curry(
  (cancel, { cwd, color: useColor, config, help, HELP, init }) => {
    const chalk = new Chalk({ level: useColor ? 2 : 0 })
    return pipe(
      cond([
        [
          K(init),
          () =>
            pipe(
              writeInitConfigFileWithCancel(cancel),
              swap
            )(path.resolve(cwd, '.gitpartyrc')),
        ],
        [K(help), K(resolve(HELP))],
        [
          T,
          pipe(
            log.configFile('trying to load config...'),
            configFileWithCancel(cancel),
            map(log.configFile('loaded...')),
            mapRej(
              pipe(
                log.configFile('error...'),
                propOr('', 'message'),
                cond([
                  [
                    includes('No config file found'),
                    K(GITPARTY_CONFIG_NEEDED(chalk)),
                  ],
                  [startsWith('SyntaxError'), K(CONFIG_NOT_VALID)],
                  [T, I],
                ])
              )
            )
          ),
        ],
      ])
    )(
      config || {
        ns: PKG.name,
      }
    )
  }
)

export const loadGitData = curry((cancel, chalk, data) =>
  pipe(
    log.config('searching for .git/index'),
    findUpWithCancel(cancel, {}),
    map(log.config('git found, path...')),
    map(path.dirname),
    chain(repo =>
      repo
        ? gitlogWithCancel(cancel, { repo })
        : rejectF(THIS_IS_NOT_A_GIT_REPO(chalk))
    ),
    map(pipe(objOf('gitlog'), mergeRight(data)))
  )(['.git/index'])
)

const adjustRelativeTimezone = curry((timeZone, preferredFormat, commit) => {
  const { authorDate } = commit
  const newDate = pipe(
    // 2023-11-21 15:58:44 -0800
    d => parseTime(d, 'yyyy-MM-dd HH:mm:ss XX', new Date()),
    timeZone.toLowerCase() === 'utc'
      ? zonedTimeToUtc
      : pipe(zonedTimeToUtc, x => utcToZonedTime(x, timeZone)),
    x => formatDate(x, preferredFormat, { timeZone })
  )(authorDate)
  commit.formattedDate = newDate
  return commit
})
const deriveAuthor = curry((lookup, commit) => {})

const getFiletype = z => z.slice(z.indexOf('.'), Infinity)

const getFiletypes = commit =>
  pipe(
    propOr([], 'files'),
    map(getFiletype),
    uniq,
    objOf('filetypes'),
    mergeRight(commit)
  )(commit)

export const processData = curry((chalk, config, data) => {
  return pipe(
    config.excludeMergeCommits
      ? reject(pipe(propOr('', 'subject'), startsWith('Merge')))
      : I,
    map(
      pipe(
        adjustRelativeTimezone(config.timezone, config.dateFormat),
        getFiletypes,
        commit =>
          pipe(
            x => [x],
            ap([propOr([], 'status'), propOr([], 'files')]),
            ([a, z]) => zip(a, z),
            groupBy(head),
            objOf('changes'),
            mergeRight(commit),
            z => mergeRight(z, { statuses: uniq(z.status) })
          )(commit)
        // applyPatternsWithChalk(chalk, config.patterns.matches
      )
    )
  )(data)
})

export const printData = curry((chalk, partyFile, config, data) =>
  pipe(
    groupBy(pipe(prop('authorDate'), split(' '), head)),
    map(
      pipe(
        map(commit => {
          const {
            statuses,
            filetypes,
            subject,
            authorName,
            abbrevHash,
            formattedDate,
          } = commit
          const matches = applyPatternsWithChalk(
            chalk,
            partyFile.patterns,
            commit
          )
          return `\n├─┬────────────────────────────┐\n│ │   ${chalk.red(
            authorName
          )} @ ${formattedDate} [${chalk.yellow(
            abbrevHash
          )}]  │\n│ │   ${subject}        │  \n│ ╰──${matches}───────────╯\n│`
        }),
        join('\n│')
      )
    ),
    toPairs,
    map(([k, v]) => chalk.inverse(' ' + k + ' ') + '\n│' + v),
    join('\n')
  )(data)
)

export const runner = curry((cancel, argv) => {
  let canon
  return pipe(
    v => {
      const cwd = process.cwd()
      return configurate(
        YARGS_CONFIG,
        { ...CONFIG_DEFAULTS, repo: cwd, cwd },
        HELP_CONFIG,
        {
          name: $NAME,
          description: $DESC,
        },
        v
      )
    },
    map(log.config('parsed args...')),
    chain(config => {
      // setup chalk
      const chalk = new Chalk({ level: config.color ? 2 : 0 })
      return pipe(
        loadPartyFile(cancel),
        // mash the data together
        map(partyFile => ({ config, partyFile, chalk })),
        chain(loadGitData(cancel, chalk))
      )(config)
    }),
    map(({ config, partyFile, gitlog, chalk }) => {
      canon = canonicalize({})
      return pipe(
        propOr({}, 'patterns'),
        map(v =>
          v?.matches
            ? mergeRight(v, {
                fn: Array.isArray(v.color)
                  ? pipe(
                      map(c => chalk[c]),
                      x => raw => pipe.apply(null, x)(raw)
                    )(v.color)
                  : chalk[v.color],
                matches: v.matches,
              })
            : v
        ),
        // z => [z],
        patterns =>
          pipe(printLegend(chalk), legend =>
            pipe(
              processData(chalk, config),
              printData(chalk, { ...partyFile, patterns }, config),
              z => legend + '\n' + z
            )(gitlog)
          )(patterns)
        // ap([printLegend(chalk), processData(chalk, config)])
      )(partyFile)
    })
  )(argv)
})
