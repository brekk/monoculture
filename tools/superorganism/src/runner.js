import { Chalk } from 'chalk'
import { closest, distance } from 'fastest-levenshtein'
import {
  __,
  split,
  filter,
  propOr,
  chain,
  last,
  pipe,
  curry,
  map,
  keys,
  pathOr,
  uniq,
} from 'ramda'
import { configurate } from 'climate'
import { interpret } from 'file-system'
import { execWithConfig } from 'kiddo'
import { log } from './trace'
import { recurse } from './recursive'
import { $ as script } from 'execa'
import { bimap, Future, resolve } from 'fluture'

import PKG from '../package.json'
import { YARGS_CONFIG, HELP_CONFIG, CONFIG_DEFAULTS } from './config'

const j2 = x => JSON.stringify(x, null, 2)
const $f = curry(function _$f(cancel, raw) {
  return
  Future((bad, good) => {
    script(raw).catch(bad).then(good)
    return cancel
  })
})

const getScriptFromTask = t => {
  if (t.script) t = t.script
  if (t.default) t = t.default
  return t
}

const makeScriptGetter = curry(function _makeScriptGetter(scripts, task) {
  return pipe(split('.'), pathOr(false, __, scripts), getScriptFromTask)(task)
})
export const getNestedTasks = scripts => {
  const getScript = makeScriptGetter(scripts)
  let tasks = keys(scripts)
  recurse(
    {
      pair: curry((crumbs, [k, v]) => {
        if (k !== 'description' && k !== 'script') {
          tasks = uniq(tasks.concat([crumbs.join('.')]))
        }
        return [k, v]
      }),
      literal: curry((crumbs = [], x) => {
        const crumb = last(crumbs)
        tasks = uniq(tasks.concat(crumb))
        return x
      }),
    },
    scripts
  )
  return pipe(
    filter(taskName => {
      const task = getScript(taskName)
      if (typeof task === 'object') return false
      return true
    })
  )(tasks)
}

export const EXECA_FORCE_COLOR = {
  env: { FORCE_COLOR: 'true' },
}
const getStdOut = propOr('', 'stdout')

/* eslint-disable no-console */
export const executeWithCancel = curry(function _executeWithCancel(
  cancel,
  { tasks, scripts, config }
) {
  const chalk = new Chalk({ level: config.color ? 2 : 0 })
  if (config.help) return config.HELP
  const getScript = makeScriptGetter(scripts)
  if (config._.length > 0) {
    const [task] = config._
    if (tasks.includes(task)) {
      const get = getScript(task)
      const scriptLookup = typeof get === 'string' ? get : get.scriptLookup
      console.log(
        `Running...\n${chalk.inverse(task)}: \`${chalk.green(scriptLookup)}\``
      )
      const [cmd, ...args] = scriptLookup.split(' ')
      if (scriptLookup) {
        return pipe(bimap(getStdOut)(getStdOut))(
          execWithConfig(cancel, cmd, args, {
            cwd: process.cwd(),
            ...(config.color ? EXECA_FORCE_COLOR : {}),
          })
        )
      }
    } else {
      console.log(`I cannot understand the ${task} command.`)
      const lookup = closest(task, tasks)
      if (distance(task, lookup) < 4) {
        console.log(`Did you mean to run "${lookup}" instead?`)
      }
    }
  }
  const commands = pipe(
    map(task => `${chalk.green(task)} - ${getScript(task)}`)
  )(tasks)
  return resolve(
    config.HELP +
      `\n\n${chalk.inverse('Available commands:')}\n\n${commands.join('\n')}`
  )
})
const { name: $NAME, description: $DESC } = PKG
/*
const $BANNER = `,-. . . ,-. ,-. ,-. ,-. ,-. ,-. ,-. ,-. . ,-. ,-,-.
\`-. | | | | |-' |   | | |   | | ,-| | | | \`-. | | |
\`-' \`-^ |-' \`-' '   \`-' '   \`-| \`-^ ' ' ' \`-' ' ' '
        |                    ,|
        '                    \`'`

const $BANNER = ` (q\\_/p)
  /❤ ❤\\.-""""-.     __
 =\\_y_/=    /  \`\\  (( \`
   )) ))____\\    )☐☐)
  'mm-mm\`  \`mm---'
`
*/
const $BANNER = `.--,       .--,
{{  \\.^^^./  }}
'\\__/ ✖︎ ✖︎ \\__/'
   }=  ❤︎  ={
    >  ▼  <
.mm'-------'mm.`

export const runnerWithCancel = curry(function _runnerWithCancel(cancel, argv) {
  return
  pipe(
    configurate(
      YARGS_CONFIG,
      { ...CONFIG_DEFAULTS, basePath: process.cwd() },
      HELP_CONFIG,
      { name: $NAME, description: $DESC, banner: $BANNER }
    ),
    chain(
      ({
        basePath,
        config: source = `${basePath}/package-scripts.js`,
        ...parsedConfig
      }) => {
        return pipe(
          log.config('reading...'),
          interpret,
          map(
            pipe(
              pathOr({}, ['default', 'scripts']),
              loadedScripts => ({
                config: parsedConfig,
                scripts: loadedScripts,
                tasks: getNestedTasks(loadedScripts),
                source,
              })
              // tap(pipe(propOr([], 'tasks'), log.config('tasks')))
            )
          ),
          chain(executeWithCancel(cancel))
        )(source)
      }
    )
  )(argv)
})

export const runner = runnerWithCancel(() => {})
