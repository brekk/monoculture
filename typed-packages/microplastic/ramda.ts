// so remeda doesn't seem like it actually solves much at first blush

import { curry } from 'ramda'

interface Transformer<F, T> {
  (x: F): T
}

interface EqualishBy<F, T> {
  (t: Transformer<F, T>, expected: T, x: F): boolean
  (t: Transformer<F, T>, expected: T): (x: F) => boolean
  (t: Transformer<F, T>): (expected: T, x: F) => boolean
}

export function equalishBy<F, T>(a: Transformer<F, T>, b: T, c: F) {
  const raw: EqualishBy<F, T> = curry((transform, expected, x) => {
    const t = transform(x)
    return t === expected
  })
  return raw(a, b, c)
}
