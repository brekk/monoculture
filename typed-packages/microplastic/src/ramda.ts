import { curry } from 'ramda'

interface Transformer<F, T> {
  (arg: F): T
}

export interface EqualishBy<F, T> {
  (transform: Transformer<F, T>, expected: T, x: F): boolean
  (transform: Transformer<F, T>, expected: T): (x: F) => boolean
  (transform: Transformer<F, T>): (expected: T, x: F) => boolean
}

export const equalishBy = curry(
  <F, T>(transform: Transformer<F, T>, expected: T, x: F) =>
    transform(x) === expected
)
