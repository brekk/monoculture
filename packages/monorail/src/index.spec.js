import { fork, resolve } from 'fluture'
import { reduce, trim, pipe, map } from 'ramda'

import {
  EXPECTED_KEYS,
  noExtraKeys,
  validatePlugins,
  testPlugin,
  checkPlugin,
} from './validate'
import { futureFileProcessor } from './runner'

test('EXPECTED_KEYS', () =>
  expect(EXPECTED_KEYS).toEqual([
    'name',
    'fn',
    'selector',
    'preserveOffset',
    'preserveLine',
    'store',
    'dependencies',
  ]))

test('noExtraKeys', () => {
  const outcome = noExtraKeys({ a: 'a', b: 'b', c: 'c' })
  expect(outcome).toEqual(`Found additional or misspelled keys: [a, b, c]`)
})

const makeCounter = () => {
  let count = 0
  return () => {
    count += 1
    return count
  }
}

test.skip('testPlugin - basics', () => {
  const counter = makeCounter()
  const myPlugin = {
    name: 'yo',
    fn: counter,
    squibble: 'zorp',
  }

  const results = testPlugin(myPlugin)
  expect(results).toEqual({
    name: true,
    preserveLine: true,
    preserveOffset: true,
    error: 'Found additional or misspelled keys: [squibble]',
    fn: true,
    store: true,
    dependencies: true,
    selector: true,
  })
})

test.skip('testPlugin - invalid', () => {
  const results = map(testPlugin)([
    {
      name: 'yo',
      fn: 'hey now',
    },
    // nothing set
    {},
    // defaults are wrong kind
    { dependencies: 1, selector: [], name: 'again', fn: () => {} },
    // misspelled keys
    {
      depepepependencies: [],
      name: 'incorrectkeys',
      fn: () => {},
      preserveLine: 1000,
    },
  ])
  expect(results).toEqual([
    {
      dependencies: true,
      fn: false,
      name: true,
      preserveLine: true,
      preserveOffset: true,
      selector: true,
      store: true,
    },
    {
      dependencies: true,
      fn: false,
      name: false,
      preserveLine: true,
      preserveOffset: true,
      selector: true,
      store: true,
    },
    {
      dependencies: false,
      fn: true,
      name: true,
      preserveLine: true,
      preserveOffset: true,
      selector: false,
      store: true,
    },
    {
      dependencies: true,
      error: 'Found additional or misspelled keys: [depepepependencies]',
      fn: true,
      name: true,
      preserveLine: false,
      preserveOffset: true,
      selector: true,
      store: true,
    },
  ])
})

test.skip('validatePlugins - basics', () => {
  const counter = makeCounter()
  const myPlugin = {
    name: 'yo',
    fn: counter,
  }

  const results = validatePlugins([myPlugin])
  expect(results).toEqual({
    correct: ['yo'],
  })
})
test.skip('validatePlugins - invalid', () => {
  const results = validatePlugins([
    { name: 'oh-cool', fn: () => {} },
    { name: 'nice', fn: () => {} },
    {
      name: 'yo',
      fn: 'hey now',
    },
    // nothing set
    {},
    // defaults are wrong kind
    { dependencies: 1, selector: [], name: 'incorrecto', fn: () => {} },
  ])
  expect(results).toEqual({
    correct: ['oh-cool', 'nice'],
    incorrect: [
      ['yo', ['fn']],
      ['unnamed', ['name', 'fn']],
      ['incorrecto', ['selector', 'dependencies']],
    ],
  })
})
test.skip('checkPlugin', () => {
  const count = makeCounter()
  const results = map(checkPlugin)([
    { name: 'cool', fn: count },
    { name: 'really cool', fn: count, dependencies: ['cool'] },
    {
      name: 'yo',
      fn: 'hey now',
    },
    {},
    { dependencies: 1, selector: [], name: 'yo', fn: () => {} },
  ])
  expect(results).toEqual([true, true, false, false, false])
})

const PLUGINS = [
  {
    name: 'wordcount',
    dependencies: [],
    fn: (c, f) =>
      pipe(
        map(([, lineContent]) => lineContent.split(' ').map(trim).length),
        reduce((a, b) => a + b, 0)
      )(f.body),
  },
  {
    default: {
      name: 't-words',
      dependencies: [],
      preserveLine: true,
      fn: (c, line) => {
        const out = line.split(' ').filter(z => z.startsWith('t')).length
        // eslint-disable-next-line no-console
        // console.log('....', c, line, out)
        return out
      },
    },
  },
]
const FILES = [
  {
    name: '/a/b/c/cool.txt',
    hash: '/a/b/c/cool.txt',
    body: `oh yeah
this is pretty
terribly
great`
      .split('\n')
      .map((x, i) => [i, x]),
  },
  {
    name: '/a/b/c/really-cool.txt',
    hash: '/a/b/c/really-cool.txt',
    body: `hey there,
this is a cool and tricky tool which can process files and answer introspection questions


newlines are also cool
`
      .split('\n')
      .map((x, i) => [i, x]),
  },
]

const EXPECTED_PROCESSING_OUTCOME = {
  events: ['t-words', 'wordcount'],
  hashMap: {
    '/a/b/c/cool.txt': '/a/b/c/cool.txt',
    '/a/b/c/really-cool.txt': '/a/b/c/really-cool.txt',
  },
  state: {
    't-words': [
      [
        '/a/b/c/cool.txt',
        {
          file: '/a/b/c/cool.txt',
          hash: '/a/b/c/cool.txt',
          body: [
            [0, 0],
            [1, 1],
            [2, 1],
            [3, 0],
          ],
        },
      ],
      [
        '/a/b/c/really-cool.txt',
        {
          file: '/a/b/c/really-cool.txt',
          hash: '/a/b/c/really-cool.txt',
          body: [
            [0, 1],
            [1, 3],
            [2, 0],
            [3, 0],
            [4, 0],
            [5, 0],
          ],
        },
      ],
    ],
    wordcount: [
      ['/a/b/c/cool.txt', 7],
      ['/a/b/c/really-cool.txt', 24],
    ],
  },
}

test('futureFileProcessor', done => {
  const ctx = {}
  const outF = futureFileProcessor(ctx, resolve(PLUGINS), resolve(FILES))
  fork(done)(out => {
    expect(out.state).toMatchSnapshot()
    done()
  })(outF)
})
