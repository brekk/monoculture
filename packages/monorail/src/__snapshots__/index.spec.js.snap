// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`futureFileProcessor 1`] = `
{
  "t-words": {
    "/a/b/c/cool.txt": {
      "body": [
        [
          0,
          0,
        ],
        [
          1,
          1,
        ],
        [
          2,
          1,
        ],
        [
          3,
          0,
        ],
      ],
      "file": "/a/b/c/cool.txt",
      "hash": "/a/b/c/cool.txt",
    },
    "/a/b/c/really-cool.txt": {
      "body": [
        [
          0,
          1,
        ],
        [
          1,
          3,
        ],
        [
          2,
          0,
        ],
        [
          3,
          0,
        ],
        [
          4,
          0,
        ],
        [
          5,
          0,
        ],
      ],
      "file": "/a/b/c/really-cool.txt",
      "hash": "/a/b/c/really-cool.txt",
    },
  },
  "wordcount": {
    "/a/b/c/cool.txt": 7,
    "/a/b/c/really-cool.txt": 24,
  },
}
`;
